fossunited/payments/doctype/razorpay_payment/test_razorpay_payment.py

```diff
@@ -1,129 +1,94 @@
-            }
-        )
+            frappe.delete_doc(RAZORPAY_PAYMENT, payment.name, force=True)
+
+        tickets = frappe.get_all(EVENT_TICKET, {"event": self.event})
+        for ticket in tickets:
+            frappe.delete_doc(EVENT_TICKET, ticket.name)
+
+        self.chapter.delete(force=True)
+        self.event.delete(force=True)
 
-        try:
-            invoice.insert()
-            frappe.db.commit()
-            return invoice
-        except Exception:
-            frappe.db.rollback()
-            raise
-
-    def test_admin_can_create_invoice(self):
-        invoice = self.create_payment_document(self.admin_user)
-        self.assertIsNotNone(invoice)
-        self.assertEqual(invoice.doctype, RAZORPAY_PAYMENT)
-        self.assertEqual(invoice.owner, self.admin_user.name)
-
-    def test_normal_user_cannot_create_invoice(self):
+    def test_payment_creation(self):
+        # Given, as administrator
+        frappe.set_user("Administrator")
+        # When a payment doctype is created
+        # Then it should be created without any errors
+        payment = insert_test_razorpay_payment(event=self.event.name)
+        self.assertTrue(payment)
+        payment.delete(force=True)
+
+        # When a non system user tries to create a payment,
+        # Then it should throw error
+        frappe.set_user("test1@example.com")
+        with self.assertRaises(frappe.PermissionError):
+            insert_test_razorpay_payment(event=self.event.name)
+
+    def test_ticket_creation_on_capture(self):
+        # Given a paid event
+        # When a payment doctype linked to this event is created
+        payment = insert_test_razorpay_payment(event=self.event.name)
+        self.assertTrue(payment)
+        # Then the payment's status should be Pending
+        self.assertEqual(payment.status, "Pending")
+        # And there should be no ticket linked to this payment doctype
+        self.assertIsNone(frappe.db.exists(EVENT_TICKET, {"razorpay_payment": payment.name}))
+
+        # When the payment status is changed to "Captured"
+        payment.status = "Captured"
+        payment.save()
+        payment.reload()
+
+        # Then a ticket should be created
+        self.assertIsNotNone(frappe.db.exists(EVENT_TICKET, {"razorpay_payment": payment.name}))
+
+    def test_payment_creation_on_closed_tickets(self):
+        # Given an event with tickets closed
+        self.event.tickets_status = "Closed"
+        self.event.save()
+        self.event.reload()
+
+        # When a payment linked to this event is created
+        payment = insert_test_razorpay_payment(event=self.event.name)
+
+        # Then it should throw an error on attempt to change its status to Captured
         with self.assertRaises(frappe.PermissionError):
-            self.create_payment_document(self.normal_user)
+            payment.status = "Captured"
+            payment.save()
+
+        payment.delete(force=True)
```

Comment:
This should be take care off in tearDown

Reply:
I think testing that on Razorpay Payment capture would make sense, adding multiple ticket creation test there

Reply:
Surprisingly, in some cases the teardown fails to remove the instances, and errors happen on other tests. This is one of the reason I prefer to just manually delete when I can.

I shall be making an issue in Framework repo for that.